File: src/database/supabase_updater.ts
// Supabase Updater em Deno
// Substitui a funcionalidade do supabase_updater.py

import { createClient } from "https://esm.sh/@supabase/supabase-js@2.39.3";

interface UpdateResult {
  success: boolean;
  action?: string;
  error?: string;
  warning?: string;
}

export class SupabaseUpdater {
  private supabase: any;
  private supabaseUrl: string;
  private supabaseKey: string;

  /**
   * Inicializa a conexão com o Supabase
   */
  constructor() {
    // Verificar se as variáveis de ambiente necessárias estão disponíveis
    this.supabaseUrl = Deno.env.get("SUPABASE_URL") || "";
    this.supabaseKey = Deno.env.get("SUPABASE_KEY") || "";

    if (!this.supabaseUrl || !this.supabaseKey) {
      throw new Error("As variáveis de ambiente SUPABASE_URL e SUPABASE_KEY são necessárias");
    }

    // Criar cliente Supabase
    try {
      this.supabase = createClient(this.supabaseUrl, this.supabaseKey);
      console.log("Cliente Supabase inicializado com sucesso (Deno)");

      // Verificar se a tabela logs existe
      this._ensureLogsTableExists();
    } catch (e) {
      console.error(`Erro ao conectar ao Supabase: ${e.message}`);
      throw e;
    }
  }

  /**
   * Verifica se a tabela logs existe, sem tentar criar
   */
  private async _ensureLogsTableExists(): Promise<boolean> {
    try {
      // Verificar se tabela existe consultando por um registro
      await this.supabase.from('logs').select('id').limit(1);
      return true;
    } catch (e) {
      console.error(`Tabela 'logs' pode não existir: ${e.message}`);
      console.log("A tabela 'logs' deve existir previamente no banco de dados.");
      console.log("Verifique a estrutura da tabela conforme documentação no arquivo config/create_logs_table.sql");
      return false;
    }
  }

  /**
   * Obtém todos os room_ids da tabela 'rooms'
   * @returns Array de room_ids
   */
  async getAllRoomIds(): Promise<string[]> {
    try {
      const { data, error } = await this.supabase
        .from('rooms')
        .select('room_id');

      if (error) {
        console.error(`Erro ao consultar room_ids: ${error.message}`);
        return [];
      }

      if (!data || data.length === 0) {
        console.log("Nenhum room_id encontrado na tabela rooms");
        return [];
      }

      return data.map(item => item.room_id.toString());
    } catch (e) {
      console.error(`Erro ao obter room_ids: ${e.message}`);
      return [];
    }
  }

  /**
   * Atualiza o título (label) de um quarto no Supabase.
   * @param roomId ID do quarto/anúncio do Airbnb
   * @param label Título do anúncio
   * @returns Resultado da operação
   */
  async updateRoomLabel(roomId: string, label: string): Promise<UpdateResult> {
    if (!roomId || !label) {
      return { success: false, error: "room_id e label são obrigatórios" };
    }

    try {
      // Primeiro verificar se o quarto já existe
      const exists = await this.roomExists(roomId);

      let result;
      let action;

      if (exists) {
        console.log(`Atualizando quarto existente ${roomId} com título: '${label}'`);
        // Atualizar quarto existente
        result = await this.supabase
          .from('rooms')
          .update({ label })
          .eq('room_id', roomId);
        action = "atualizado";
      } else {
        console.log(`Inserindo novo quarto ${roomId} com título: '${label}'`);
        // Inserir novo quarto
        result = await this.supabase
          .from('rooms')
          .insert({ room_id: roomId, label });
        action = "inserido";
      }

      // Verificar se houve erro
      if (result.error) {
        console.error(`Erro ao ${action} no Supabase: ${result.error.message}`);
        return { success: false, error: result.error.message };
      }

      console.log(`✅ Quarto ${roomId} ${action} com sucesso no Supabase (Deno)`);
      return { success: true, action };
    } catch (e) {
      console.error(`❌ Falha ao atualizar o quarto ${roomId}: ${e.message}`);
      return { success: false, error: e.message };
    }
  }

  /**
   * Verifica se um quarto existe na tabela rooms
   * @param roomId ID do quarto/anúncio do Airbnb
   * @returns True se o quarto existir
   */
  async roomExists(roomId: string): Promise<boolean> {
    try {
      const { data, error } = await this.supabase
        .from('rooms')
        .select('room_id')
        .eq('room_id', roomId);

      if (error) {
        console.error(`Erro ao verificar quarto: ${error.message}`);
        return false;
      }

      return data && data.length > 0;
    } catch (e) {
      console.error(`Falha ao verificar quarto ${roomId}: ${e.message}`);
      return false;
    }
  }

  /**
   * Registra uma execução do scraper na tabela logs (opcional)
   * Essa função é totalmente opcional e não afeta o funcionamento principal do sistema.
   * A tabela 'logs' deve existir previamente no banco de dados com a estrutura:
   * - id UUID PRIMARY KEY
   * - service_time FLOAT NOT NULL
   * - room_ids INTEGER[] NOT NULL
   * - success_count INTEGER NOT NULL
   * - total_count INTEGER NOT NULL
   * - created_at TIMESTAMP WITH TIME ZONE
   * 
   * @param serviceTime Tempo de execução em segundos
   * @param roomIds Lista de IDs dos quartos processados
   * @param successCount Número de quartos processados com sucesso
   * @param totalCount Número total de quartos processados
   * @returns Resultado da operação
   */
  async logExecution(
    serviceTime: number,
    roomIds: number[],
    successCount: number,
    totalCount: number
  ): Promise<UpdateResult> {
    try {
      console.log(`Registrando log de execução no Supabase (opcional):`);
      console.log(`  Tempo de execução: ${serviceTime.toFixed(2)} segundos`);
      console.log(`  Quartos processados: ${roomIds.length}`);
      const successRate = totalCount > 0 ? (successCount / totalCount * 100).toFixed(1) : "0.0";
      console.log(`  Taxa de sucesso: ${successCount}/${totalCount} (${successRate}%)`);

      // Verificar se podemos fazer log (a tabela pode não existir)
      try {
        // Tentando inserir com a estrutura conforme config/create_logs_table.sql
        const insertData = {
          service_time: serviceTime,
          room_ids: roomIds,
          success_count: successCount,
          total_count: totalCount,
          created_at: new Date().toISOString()
        };
        
        // Tentativa silenciosa - se falhar, apenas ignoramos sem mostrar erros
        const { error } = await this.supabase
          .from('logs')
          .insert(insertData);

        if (error) {
          // Falha silenciosa - apenas registramos no console para debug, mas não geramos erro
          if (error.message.includes("Could not find")) {
            console.log("ℹ️ A tabela 'logs' pode não existir no Supabase ou ter uma estrutura diferente.");
            console.log("ℹ️ Esta funcionalidade é opcional e não afeta o funcionamento principal.");
          } else {
            console.log(`ℹ️ Log opcional não registrado: ${error.message.split('\n')[0]}`);
          }
          // Continuamos o processo normalmente
          return { success: true };
        }

        console.log(`✅ Log de execução registrado com sucesso`);
        return { success: true };
      } catch (logError) {
        // Silenciamos qualquer erro do log, pois é funcionalidade opcional
        console.log(`ℹ️ O registro de log é opcional e não afeta a operação principal`);
        return { success: true };
      }
    } catch (e) {
      // Silenciamos qualquer erro do log, pois é funcionalidade opcional
      return { success: true };
    }
  }
}

/**
 * Função para atualizar o título a partir da linha de comando
 * Uso: deno run --allow-net --allow-env supabase_updater.ts ROOM_ID "Título do anúncio"
 */
async function updateFromCommandLine(): Promise<void> {
  const args = Deno.args;
  
  if (args.length < 2) {
    console.log("Uso: deno run --allow-net --allow-env supabase_updater.ts ROOM_ID 'Título do anúncio'");
    Deno.exit(1);
  }

  const roomId = args[0];
  const label = args[1];

  try {
    const updater = new SupabaseUpdater();
    const result = await updater.updateRoomLabel(roomId, label);

    if (result.success) {
      console.log(`Título atualizado com sucesso para o quarto ${roomId}`);
      Deno.exit(0);
    } else {
      console.error(`Falha ao atualizar o título: ${result.error || 'Erro desconhecido'}`);
      Deno.exit(1);
    }
  } catch (e) {
    console.error(`Erro ao executar a atualização: ${e.message}`);
    Deno.exit(1);
  }
}

// Se executado diretamente, chamar a função de atualização
if (import.meta.main) {
  updateFromCommandLine();
}

File: supabase_updater.py
#!/usr/bin/env python3
"""
Módulo para atualizar títulos de anúncios do Airbnb no Supabase.
"""

import os
import sys
import time
from supabase import create_client

class SupabaseUpdater:
    """
    Classe para gerenciar a conexão e atualizações no Supabase
    """
    
    def __init__(self):
        """
        Inicializa a conexão com o Supabase
        """
        # Verificar se as variáveis de ambiente necessárias estão disponíveis
        self.supabase_url = os.environ.get('SUPABASE_URL')
        self.supabase_key = os.environ.get('SUPABASE_KEY')
        
        if not self.supabase_url or not self.supabase_key:
            raise ValueError("As variáveis de ambiente SUPABASE_URL e SUPABASE_KEY são necessárias")
        
        # Criar cliente Supabase
        try:
            self.supabase = create_client(self.supabase_url, self.supabase_key)
            print("Cliente Supabase inicializado com sucesso (Python)")
            
            # Verificar se a tabela logs existe
            self._ensure_logs_table_exists()
        except Exception as e:
            print(f"Erro ao conectar ao Supabase: {str(e)}")
            raise
    
    def _ensure_logs_table_exists(self):
        """
        Verifica se a tabela logs existe e cria se necessário
        """
        try:
            # Verificar se tabela existe consultando por um registro
            self.supabase.table('logs').select('id').limit(1).execute()
        except Exception as e:
            print(f"Tabela 'logs' pode não existir: {str(e)}")
            print("Consulte as instruções para criar a tabela logs no arquivo create_logs_table.sql")
    
    def update_room_label(self, room_id, label):
        """
        Atualiza o título (label) de um quarto no Supabase.
        Se o quarto não existir, ele é inserido.
        Se já existir, é atualizado.
        
        Args:
            room_id: ID do quarto/anúncio do Airbnb
            label: Título do anúncio
        
        Returns:
            dict: Resultado da operação
        """
        if not room_id or not label:
            return {"success": False, "error": "room_id e label são obrigatórios"}
        
        try:
            # Primeiro verificar se o quarto já existe
            exists = self.room_exists(room_id)
            
            if exists:
                print(f"Atualizando quarto existente {room_id} com título: '{label}'")
                # Atualizar quarto existente
                result = self.supabase.table('rooms').update(
                    {"label": label}
                ).eq('room_id', room_id).execute()
                action = "atualizado"
            else:
                print(f"Inserindo novo quarto {room_id} com título: '{label}'")
                # Inserir novo quarto
                result = self.supabase.table('rooms').insert(
                    {"room_id": room_id, "label": label}
                ).execute()
                action = "inserido"
            
            # Verificar se houve erro
            if hasattr(result, 'error') and result.error:
                print(f"Erro ao {action} no Supabase: {result.error}")
                return {"success": False, "error": str(result.error)}
            
            print(f"✅ Quarto {room_id} {action} com sucesso no Supabase (Python)")
            return {"success": True, "action": action}
            
        except Exception as e:
            print(f"❌ Falha ao atualizar o quarto {room_id}: {str(e)}")
            return {"success": False, "error": str(e)}
    
    def room_exists(self, room_id):
        """
        Verifica se um quarto existe na tabela rooms
        
        Args:
            room_id: ID do quarto/anúncio do Airbnb
        
        Returns:
            bool: True se o quarto existir
        """
        try:
            result = self.supabase.table('rooms').select('room_id').eq('room_id', room_id).execute()
            
            # Verificar se houve erro
            if hasattr(result, 'error') and result.error:
                print(f"Erro ao verificar quarto: {result.error}")
                return False
            
            # Verificar se há dados
            data = result.data if hasattr(result, 'data') else []
            
            return len(data) > 0
            
        except Exception as e:
            print(f"Falha ao verificar quarto {room_id}: {str(e)}")
            return False
            
    def log_execution(self, service_time, rooms_ids, success_count, total_count):
        """
        Registra uma execução do scraper na tabela logs
        
        Args:
            service_time: Tempo de execução em segundos
            rooms_ids: Lista de IDs dos quartos processados
            success_count: Número de quartos processados com sucesso
            total_count: Número total de quartos processados
            
        Returns:
            dict: Resultado da operação
        """
        try:
            print(f"Registrando log de execução no Supabase:")
            print(f"  Tempo de execução: {service_time:.2f} segundos")
            print(f"  Quartos processados: {len(rooms_ids)}")
            print(f"  Taxa de sucesso: {success_count}/{total_count} ({(success_count/total_count*100):.1f}%)")
            
            # Inserir registro no log
            result = self.supabase.table('logs').insert({
                "service_time": service_time,
                "room_ids": rooms_ids,  # Usando room_ids (sem o 's' final)
                "success_count": success_count,
                "total_count": total_count
            }).execute()
            
            # Verificar se houve erro
            if hasattr(result, 'error') and result.error:
                print(f"Erro ao registrar log no Supabase: {result.error}")
                return {"success": False, "error": str(result.error)}
            
            print(f"✅ Log de execução registrado com sucesso no Supabase")
            return {"success": True}
            
        except Exception as e:
            print(f"❌ Falha ao registrar log: {str(e)}")
            return {"success": False, "error": str(e)}

def update_from_commandline():
    """
    Função para atualizar o título a partir da linha de comando
    Uso: python supabase_updater.py ROOM_ID "Título do anúncio"
    """
    if len(sys.argv) < 3:
        print("Uso: python supabase_updater.py ROOM_ID 'Título do anúncio'")
        sys.exit(1)
    
    room_id = sys.argv[1]
    label = sys.argv[2]
    
    try:
        updater = SupabaseUpdater()
        result = updater.update_room_label(room_id, label)
        
        if result.get('success'):
            print(f"Título atualizado com sucesso para o quarto {room_id}")
            sys.exit(0)
        else:
            print(f"Falha ao atualizar o título: {result.get('error', 'Erro desconhecido')}")
            sys.exit(1)
    except Exception as e:
        print(f"Erro ao executar a atualização: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    update_from_commandline()